cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt

```diff
@@ -691,27 +718,60 @@ fun Node.followNextDFGEdgesUntilHit(
                             // from).
                             // We try to pop from the stack and only select the elements with the
                             // matching index.
-                            indexStack?.checkAndPop(it.granularity as IndexedDataflowGranularity) ==
-                                true
+                            ctx.indexStack.checkAndPop(
+                                it.granularity as IndexedDataflowGranularity
+                            ) == true
                         } else {
                             true
                         }
                     }
                     .map { it.end }
             } else {
-                // We don't have to care about poping stuff from the stack. Maybe we can push
-                // something but we accept everything.
+                // First, let's take care of pushing any information on the stacks (index, calling
+                // context)
                 currentNode.nextDFGEdges.forEach {
+                    if (it is ContextSensitiveDataflow && it.callingContext is CallingContextIn) {
+                        // Push the call of our calling context to the stack
+                        ctx.callStack.addFirst(it.callingContext.call)
+                    }
                     if (
                         it.end is InitializerListExpression &&
                             it.granularity is IndexedDataflowGranularity
                     ) {
                         // CurrentNode is the child and nextDFG goes to ILE => currentNode's written
                         // to. Push to stack
-                        indexStack?.push(it.granularity as IndexedDataflowGranularity)
+                        ctx.indexStack.push(it.granularity as IndexedDataflowGranularity)
                     }
                 }
-                currentNode.nextDFG
+
+                // We need to filter out the edges which based on the stack
+                val selected =
```

Comment:
We should probably make a single run over the edges based on the `ctx` and its stacks and make this context configurable and include the push/pop logic. Otherwise, we'll have many parts of code which share very similar functionality and will run into issues with the maintainability. Might be a goal for the next few weeks

Reply:
Agreed, should be part of https://github.com/Fraunhofer-AISEC/cpg/issues/1991